Biblioteca Nacional de Maestros.js

```diff
@@ -0,0 +1,211 @@
+{
+	"translatorID": "b383df35-15e7-43ee-acd9-88fd62669083",
+	"label": "Biblioteca Nacional de Maestros",
+	"creator": "Sebastian Karcher",
+	"target": "^https?://www\\.bnm\\.me\\.gov\\.ar/catalogo",
+	"minVersion": "3.0",
+	"maxVersion": "",
+	"priority": 100,
+	"inRepository": true,
+	"translatorType": 4,
+	"browserSupport": "gcsibv",
+	"lastUpdated": "2020-06-21 20:46:20"
+}
+
+/*
+	***** BEGIN LICENSE BLOCK *****
+
+	Copyright © 2020 Sebastian Karcher
+	
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+
+function detectWeb(doc, url) {
+	// We'll go with Book throughout for simplicity's sake
+	if (/\/Record\/\d+/.test(url)) {
+		return "book";
+	}
+	else if (getSearchResults(doc, true)) {
+		return "multiple";
+	}
+	return false;
+}
+
+function getSearchResults(doc, checkOnly) {
+	var items = {};
+	var found = false;
+	var rows = doc.querySelectorAll('.result div[class*="resultItemLine"]>a.title[href*="/Record/"]');
+	for (let row of rows) {
+		let href = row.href;
+		let title = ZU.trimInternal(row.textContent);
+		if (!href || !title) continue;
+		if (checkOnly) return true;
+		found = true;
+		items[href] = title;
+	}
+	return found ? items : false;
+}
+
+function doWeb(doc, url) {
+	if (detectWeb(doc, url) == "multiple") {
+		Zotero.selectItems(getSearchResults(doc, false), function (items) {
+			if (items) scrape(Object.keys(items));
+		});
+	}
+	else {
+		scrape([url]);
+	}
+}
+
+
+function constructMARCurl(url) {
+	url = url.replace(/\/(Details|Holdings)([#?].*)?/, ""); // remove panels
+	return url + "/Export?style=MARCXML";
+}
+
+
+function scrape(urls) {
+	var handler = function (text) {
+		var translator = Zotero.loadTranslator("import");
+		// Z.debug(text);
+		// MARC XML
+		translator.setTranslator("edd87d07-9194-42f8-b2ad-997c4c7deefd");
+		translator.setString(text);
+		translator.translate();
+	};
+	for (let url of urls) {
+		// Z.debug(url);
+		ZU.doGet(constructMARCurl(url), handler);
```

Comment:
Actually, there's a more fundamental problem with this. When `doGet()` is used inside a loop, it's called simultaneously as many times as there are loops, resulting in many simultaneous requests (possibly limited to 6 by the network stack). doGet() can take an array of URLs, and it will automatically serialize those rather than trying to make all the requests at once. (In the future, it (or its successor) could also gently parallelize them.) So you can just remove the loop here and call doGet() once with an array of constructed URLs, with a single function for the callback.

I suspect this mistake is made in many translators. It's actually done in the Google Scholar translator, which might be the reason for the frequent throttling we see there.

Reply:
That makes sense -- I think I got this right in https://github.com/zotero/translators/pull/2194/commits/adff8648ceca06a648b80d96a9774dad0a04be82 ?